### ISSUE TITLE
IP objects proposal

### ISSUE THREAD

**Issue by trustmaster** [noflo/noflo#290]
At present NoFlo IPs are just raw JavaScript values which is simple and easy to use once you have the data. At the same time NoFlo utilizes groups as both substream separators (brackets) and additional data properties. There are 7 types of events that occur on ports: `attach`, `detach`, `connect`, `disconnect`, `beginGroup`, `endGroup` and `data`. Connect and disconnect events are also similar to substream brackets.

Meanwhile in Classical FBP IPs are always objects of IP class, enveloping raw data. Bracket IPs are just IPs of special type which mark beginnings and endings of substreams and nothing else. There is just one event type: `ip`. In some cases there can also be attach/detach type events. But components normally deal with IPs only and decide how to handle them.

Despite the simplicity of packet contents in NoFlo, there are difficulties caused by such granular design:

* Component author has to make a decision on which type of events a component should do actual work. Mixing event types on different ports of the same component and across components is hard and error-prone.
* Groups are mostly used as IP meta tags rather than substream brackets.
* Group and data packets are decoupled from each other but are usually consumed all together. This floods the event loop with unnecessary events and creates synchronization code overhead.
* Divergent-convergent graphs are hard on synchronization.

A proposed change is to wrap data into IP envelope objects instead of sending raw data and use classical bracket IPs for substreaming purpose only. An IP object would contain the following:

 * `data`: raw data
 * `groups`: array of attached group tags
 * any additional metadata, e.g. request ID

Bracket IPs would be special IP types holding bracket type (opening or closing) instead of data.

The upsides of such a solution are:

 * No event decision headache: the only event that triggers a process is a ‘new IP’.
 * Groups are properties of a packet, no need to receive them separately and attach them to a packet manually.
 * Stream processing logic is explicit.
 * Divergent-convergent synchronization can be described with simpler attributes and predicates.

There are of course downsides:

 * Accessing raw IP data is more verbose: `ip.data` instead of just `ip`.
 * Backwards compatibility is not a given but is some work to do.

This would get NoFlo a bit closer to Classical FBP and its best practices, yet without trying to force threadful CFBP’s concurrency model on eventful ECMAScript ground. We could even implement optional behavior like exclusive IP ownership, cloning and explicit disposal later on.

Please consider this issue as a proposal, not a demand. I'd like to hear some opinions on this, especially from the browser side as I'm dealing with Node.js problems mostly.

**Comment by jonnor** [comment]
I'm generally pro this suggestion. It will make group forwarding just-work in most cases (just manipulate `ip.data`) instead of having to use some special magic (like WirePattern).

I think it is critical that we have compatibility though, at least in a transition phase. Flag-day migration don't work.

**Comment by bergie** [comment]
:+1: from me, given that we sort out the backwards compatibility question. This switch may best be coupled with redesign of the component API to move firing patterns out of helpers and into component API itself.

This way old-style components would use the old (and deprecated) API, and still get the events and data they expect, and components could be written (and ported) to new API as we go.

**Comment by trustmaster** [comment]
Yes, this is how I see the backwards compatibility: the old Ports API will emulate the old-style raw data and events.

The next question is: given 3 years of NoFlo experience by now and the IPs being switched to objects, do we still need all 7 types of events in the new API? Or can `connect`, `disconnect`, `beginGroup`, `endGroup` be safely substituted with just bracket IPs?

**Comment by bergie** [comment]
I suppose they can. `connect` and `disconnect` are essentially the outermost brackets anyway

**Comment by bergie** [comment]
We should also have a tool for recognizing components using the legacy API. Would help tracking down stuff that needs to be ported

**Comment by trustmaster** [comment]
Here is where I got so far: https://github.com/trustmaster/noflo/blob/ip-objects/spec/Component.coffee#L212

See both test cases for new Ports API (yes, sorry for that) example.

I had to use some new method names to keep the old API fully available.

The `InPort` [maintains](https://github.com/trustmaster/noflo/blob/ip-objects/spec/InPort.coffee#L263) compatibility layer between the old raw packets with 7 event types and the new object packets with 1 event type. So the old-style components can be combined with new-style components within the same graph.

I've added [`InternalSocket.post()`](https://github.com/trustmaster/noflo/blob/ip-objects/src/lib/InternalSocket.coffee#L93) method because connect/disconnect events are considered as optional. This is to avoid the annoying `disconnect()` every time you just want to send a single packet, and to avoid littering the compatibility layer with unnecessary `openBracket/closeBracket` packets.
